Adding support for copying container images with image digest [skopeo only]

Nicolas-Peiffer opened the pull request:
Adding support for copying container images with image digest [skopeo only].

## Related issue

https://github.com/xelalexv/dregsy/issues/86

## TODOs

I started implementing some tests, but tests needs to be improved.

## Feature details

- For Skopeo relay, adding support for pulling container images based on the image digest. An image digest looks like this: `sha256:5e8e0509e829bb8f990249135a36e81a3ecbe94294e7a185cc14616e5fad96bd`.
  Below is the behavior of the `skopeo copy` Sync with support for both
  digests and tags.
  Warning: Skopeo Docker references with both a tag and a digest are
  currently not supported.

  The `digests` list & `tags` list are stored in a `mapping` struct.
  As example, look at the file
  `test/fixtures/config/skopeo-digest-only-valid.yaml`

  | `digests` list | `tags` list | `dregsy` behavior                             | diff with 0.4.4 |
  |----------------|-------------|-----------------------------------------------|-----------------|
  | empty          | empty       | pulls all tags                                | same            |
  | empty          | NOT empty   | pulls filtered tags only                      | same            |
  | NOT empty      | NOT empty   | pulls filtered tags AND pulls correct digests | different       |
  | NOT empty      | empty       | pulls correct digests only, ignores tags      | different       |

  A "correct digest" is a correctly formated AND an existing digest.
  Skopeo is used to verify if the digest exists before trying to copy it.
- Adding a simple bash script that uses `openssl` and `podman` to instantiate a simple registryv2 service with an HTTPS endpoint: `test/bin/create_local_registry_https.sh`, for testing purpose.
- Adding `dregsy` configuration samples, for testing purpose:
    + `skopeo-digest-bad-formated.yaml`: a config sample with image digests that are badly formated. Dregsy should raise an error in logs and skip to next tag or digest.
    + `skopeo-digest-dont-exist.yaml`: a config sample with image digests that are formated correctly, but that does not exist. Dregsy should raise an error in logs and skip to next tag or digest.
    + `skopeo-digest-duplicates.yaml`: a config sample with duplicates image digests. Dregsy should eliminate duplicates and only copy once.
    + `skopeo-digest-only-valid.yaml`: a config sample with valid image digests that exist on docker hub (`docker.io`). Dregsy should copy them without issue.
- Adding 2 `podman` or `docker` Containerfiles for x86 and ARM64 architecture: `Containerfile.aarch64-arm64` and `Containerfile.x86_64-amd64`. For testing purpose and local build.
- Adding a section in the README: _### Building only the `dregsy` binary with `podman`_
- upgrades:
    + *Go* 1.20.2 in `Containerfile.aarch64-arm64`, `Containerfile.x86_64-amd64`.

xelalexv commented:
Thanks for your PR! I did a first review, here are my comments:

- We should focus the PR on just adding the digest functionality:
    - Do not include the Container files. Moving to Podman for building dregsy and images is something that has been on my mind for a while. When switching to Podman, this should be done fully for all build actions, and would require a separate PR. This however does not have the highest priority right now.
    - Also remove the script for creating the HTTPS enabled registry, as this is also out of focus for this PR.
- Please squash the remaining commits.

Let's continue the discussion about the implementation approach in #95.

nicop311 commented:
I mostly agree with what you say. I will use https://github.com/xelalexv/dregsy/issues/86 to discuss PR content, but I just answer some of your points here.

* _Do not include the Container files_ : I agree with you. However, I needed to do create them, because I could not use `docker` CLI and I could not use your github actions to build ant test Dregsy. But I agree to remove them, this is why I did a specific commit for this.
* _switching to `podman` does not have the highest priority right now_: this is fine. However, I think you should build the dregsy container for both amd64 x86 **and aarch64 arm64/v8** (for Raspberry Pies & co).
* _Also remove the script for creating the HTTPS enabled registry_ : again same idea for this one, I needed an environment to test `dregsy` with the new feature. If you have a better way to do this sort of test, lets use it.
* _squash commits_: once we are happy with content, we can rebase / squash into 1 commit no problem.

Nicolas-Peiffer commented:
I mostly agree with what you say. I will use https://github.com/xelalexv/dregsy/issues/86 to discuss PR content, but I just answer some of your points here.

* _Do not include the Container files_ : I agree with you. However, I needed to do create them, because I could not use `docker` CLI and I could not use your github actions to build ant test Dregsy. But I agree to remove them, this is why I did a specific commit for this.
* _switching to `podman` does not have the highest priority right now_: this is fine. However, I think you should build the dregsy container for both amd64 x86 **and aarch64 arm64/v8** (for Raspberry Pies & co).
* _Also remove the script for creating the HTTPS enabled registry_ : again same idea for this one, I needed an environment to test `dregsy` with the new feature. If you have a better way to do this sort of test, lets use it.
* _squash commits_: once we are happy with content, we can rebase / squash into 1 commit no problem.

xelalexv commented:
>  _Do not include the Container files_ : I agree with you. However, I needed to do create them, because I could not use `docker` CLI and I could not use your github actions to build ant test Dregsy.

Actually, the dregsy project does not use GitHub actions. All build & test functionality is provided by the `Makefile`, e.g. `make dregsy` or `make imgdregsy`. That however requires *Docker* in your system, which is the reason why I want to move to *Podman*.

> ...I think you should build the dregsy container for both amd64 x86 **and aarch64 arm64/v8** (for Raspberry Pies & co). 

Building via the `Makefile` already supports this. The architecture is currently picked automatically based on the host system, so when building on an ARM64 system, you get an ARM64 binary & image. What we would need to enable is cross compiling & image build.

>  _Also remove the script for creating the HTTPS enabled registry_ : again same idea for this one, I needed an environment to test `dregsy` with the new feature. If you have a better way to do this sort of test, lets use it.

This is also provided in the `Makefile` (see `registry*` targets), so for example `make tests` already sets up a local registry for testing automatically in the background.

Nicolas-Peiffer commented:
FYI:
- [x] I removed `Containerfiles` and custom bash scripts.
- [x] I removed some `auth:` tokens from my YAML config samples.
- [x] I squashed into 1 single signed commit.
- [ ] I did not modified anything else in term of the Go source code.

xelalexv commented:
closing as discussed in #86
